src/orange/univ/RectArrayTracker.hh

```diff
@@ -0,0 +1,365 @@
+    }
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate nearest distance to a surface in any direction.
+ *
+ * On an axis-aligned rectlinear grid the minimum distance to any surface is
+ * always always occurs along a line parallel to an axis.
+ */
+CELER_FUNCTION real_type RectArrayTracker::safety(Real3 const& pos,
+                                                  LocalVolumeId volid) const
+{
+    CELER_EXPECT(volid && volid.get() < this->num_volumes());
+
+    VolumeInverseIndexer vii(dims_);
+    auto coords = vii(volid.unchecked_get());
+
+    real_type min_dist = numeric_limits<real_type>::infinity();
+
+    for (auto ax : range(Axis::size_))
+    {
+        for (auto i : range(2))
+        {
+            auto target_coord = coords[to_int(ax)] + i;
+            real_type target
+                = params_.reals[record_.grid[to_int(ax)]][target_coord];
+            min_dist = min(min_dist, fabs(pos[to_int(ax)] - target));
+        }
+    }
+
+    CELER_ENSURE(min_dist >= 0
+                 && min_dist < numeric_limits<real_type>::infinity());
+    return min_dist;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate the local surface normal.
+ */
+CELER_FUNCTION auto RectArrayTracker::normal([[maybe_unused]] Real3 const& pos,
+                                             LocalSurfaceId surf) const -> Real3
+{
+    CELER_EXPECT(surf && surf.get() < this->num_surfaces());
+    SurfaceInverseIndexer sii(surface_indexer_data_);
+    size_type ax = sii(surf.get())[0];
+
+    Real3 normal{0., 0., 0.};
+    normal[ax] = 1.0;
+
+    return normal;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * DaughterId of universe embedded in a given volume.
+ */
+CELER_FORCEINLINE_FUNCTION DaughterId
+RectArrayTracker::daughter(LocalVolumeId vol) const
+{
+    CELER_EXPECT(vol && vol.get() < this->num_volumes());
+    return record_.daughters[vol];
+}
+
+//---------------------------------------------------------------------------//
+// PRIVATE INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate distance-to-intercept for the next surface.
+ */
+template<class F>
+CELER_FUNCTION auto
+RectArrayTracker::intersect_impl(LocalState const& state, F is_valid) const
+    -> Intersection
+{
+    CELER_EXPECT(state.volume && !state.temp_sense.empty());
+
+    VolumeInverseIndexer vii(dims_);
+    auto coords = vii(state.volume.unchecked_get());
+
+    Intersection result;
+    Sense sense;
+    SurfaceIndexer si(surface_indexer_data_);
+
+    for (auto ax : range(Axis::size_))
+    {
+        auto dir = state.dir[to_int(ax)];
+
+        // Ignore any stationary axis
+        if (dir == 0.0)
+        {
+            continue;
+        }
+
+        auto target_coord = coords[to_int(ax)] + static_cast<int>(dir > 0.);
```

Comment:
```suggestion
        auto target_coord = coords[to_int(ax)] + static_cast<int>(dir > 0);
```

Remember `0.` is a double literal, so I think comparisons (and certainly operations) will upcast the float.